backend/worker/autoresolver.go

```diff
@@ -0,0 +1,86 @@
+package worker
+
+import (
+	"context"
+	"time"
+
+	"github.com/highlight-run/highlight/backend/model"
+	privateModel "github.com/highlight-run/highlight/backend/private-graph/graph/model"
+	"github.com/highlight-run/highlight/backend/store"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type AutoResolver struct {
+	store *store.Store
+	db    *gorm.DB
+}
+
+func NewAutoResolver(store *store.Store, db *gorm.DB) *AutoResolver {
+	return &AutoResolver{
+		store: store,
+		db:    db,
+	}
+}
+
+func (autoResolver *AutoResolver) AutoResolveStaleErrors(ctx context.Context) {
+	projectFilterSettings := autoResolver.store.FindProjectsWithAutoResolveSetting(ctx)
+
+	for _, projectFilterSettings := range projectFilterSettings {
+		log.WithContext(ctx).WithFields(
+			log.Fields{"project_id": projectFilterSettings.ProjectID}).Info("Finding stale errors for project")
+		interval := projectFilterSettings.AutoResolveStaleErrorsDayInterval
+
+		project, err := autoResolver.store.GetProject(ctx, projectFilterSettings.ProjectID)
+		if err != nil {
+			log.WithContext(ctx).WithFields(log.Fields{"project_id": projectFilterSettings.ProjectID}).Error(err)
+			continue
+		}
+
+		err = autoResolver.resolveStaleErrorsForProjectInBatches(ctx, project, interval)
+
+		if err != nil {
+			log.WithContext(ctx).WithFields(log.Fields{"project_id": projectFilterSettings.ProjectID}).Error(err)
+			continue
+		}
+	}
+}
+
+func (autoResolver *AutoResolver) resolveStaleErrorsForProjectInBatches(ctx context.Context, project model.Project, interval int) error {
+	var errorGroups []model.ErrorGroup
+
+	db := autoResolver.db
+
+	subQuery := db.
+		Model(model.ErrorObject{}).
+		Select("error_group_id").
+		Where("created_at >= ?", time.Now().AddDate(0, 0, -interval))
+
+	result := db.Debug().Where(model.ErrorGroup{
+		State:     privateModel.ErrorStateOpen,
+		ProjectID: project.ID,
+	}).
+		Where("id NOT IN (?)", subQuery).
+		FindInBatches(&errorGroups, 100, func(tx *gorm.DB, batch int) error {
```

Comment:
Using [`FindInBatches`](https://gorm.io/docs/advanced_query.html#FindInBatches) such that for projects with lots of stale errors, we don't timeout the query.

For example, to get the last month of our project's stale errors:

```sql
SELECT
	*
FROM
	"error_groups"
WHERE
	"error_groups"."project_id" = 1
	AND "error_groups"."state" = 'OPEN'
	AND id NOT IN(
		SELECT
			"error_group_id" FROM "error_objects"
		WHERE
			created_at >= '2023-04-18 11:29:53.374')
ORDER BY
	"error_groups"."id"
```

takes 221.621s and returns 128,079 rows

Using `FindInBatches` will attach a `LIMIT 100` to the query and work through 100 error groups before fetching the next batch.

Reply:
i would avoid using `FindInBatches` here since postgres queries cannot time out but this adds some extra complexity. this will also mean that individual updates will not happen atomically since they will be part of different transactions.